### PR TITLE
Feature/lims 1760 arimport create batch

### PR DESCRIPTION
**Pull request by rockfruit**

### CODE REVIEW

**Comment by rockfruit** [comment]
Running bika.lims.testing.BikaSimple:Functional tests:
  Set up plone.testing.zca.LayerCleanup in 0.000 seconds.
  Set up plone.testing.z2.Startup in 0.192 seconds.
  Set up plone.app.testing.layers.PloneFixture in 6.657 seconds.
  Set up bika.lims.testing.SimpleTestLayer in 10.973 seconds.
  Set up bika.lims.testing.BikaSimple:Functional in 0.000 seconds.
  Running:
                
  Ran 1 tests with 0 failures and 0 errors in 7.847 seconds.
Tearing down left over layers:
  Tear down bika.lims.testing.BikaSimple:Functional in 0.000 seconds.
  Tear down bika.lims.testing.SimpleTestLayer in 0.006 seconds.
  Tear down plone.app.testing.layers.PloneFixture in 0.056 seconds.
  Tear down plone.testing.z2.Startup in 0.005 seconds.
  Tear down plone.testing.zca.LayerCleanup in 0.003 seconds.

File: bika/lims/workflow.py [hunk 1]
```diff
@@ -51,6 +52,19 @@ def doActionFor(instance, action_id):
...
 
 
+def BeforeTransitionEventHandler(instance, event):
+    """This will run the workflow_before_* on any
+    content type that has one.
+    """
+    # creation doesn't have a 'transition'
+    if not event.transition:
+        return
+    key = 'workflow_before_' + event.transition.id
```
Comment (rockfruit) [inline bika/lims/workflow.py]: now we have workflow_before_<action> for when some transition must perhaps be avoided, without aborting the transaction.

File: setup.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def read(*rnames):
           'plone.jsonapi.core',
           'Products.ATExtensions>=1.1a3',
           'Products.CMFEditions',
+          'Products.DataGridField',
```
Comment (rockfruit) [inline setup.py]: Products.DataGridField is much nicer than RecordsField and RecordsWidget.

**Comment by xispa** [comment]
Pheew.. a lot of work here! 3.1.9 is going to be special!

I'm probably missing something. I've tried with the test file, but I get this:
```
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module Shared.DC.Scripts.Bindings, line 322, in __call__
  Module Shared.DC.Scripts.Bindings, line 359, in _bindAndExec
  Module Products.CMFCore.FSPageTemplate, line 237, in _exec
  Module Products.CMFCore.FSPageTemplate, line 177, in pt_render
  Module Products.PageTemplates.PageTemplate, line 79, in pt_render
  Module zope.pagetemplate.pagetemplate, line 132, in pt_render
   - Warning: Macro expansion failed
   - Warning: <type 'exceptions.KeyError'>: 'macro'
  Module zope.pagetemplate.pagetemplate, line 240, in __call__
  Module zope.tal.talinterpreter, line 271, in __call__
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 888, in do_useMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 954, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 858, in do_defineMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 954, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 954, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 946, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 858, in do_defineMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 888, in do_useMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 821, in do_loop_tal
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 852, in do_condition
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 888, in do_useMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 888, in do_useMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 583, in do_setLocal_tal
  Module zope.tales.tales, line 696, in evaluate
   - URL: file:/home/jordi/dev/xispa/bikalims/buildout-cache/eggs/Products.Archetypes-1.9.10-py2.7.egg/Products/Archetypes/skins/archetypes/widgets/field.pt
   - Line 12, Column 4
   - Expression: <PythonExpr context.widget(field.getName(), mode=mode, use_label=1)>
   - Names:
      {'container': <PloneSite at /Plone>,
       'context': <ARImport at /Plone/clients/client-1/AI15001>,
       'default': <object object at 0x7f6485fb3b60>,
       'here': <ARImport at /Plone/clients/client-1/AI15001>,
       'loop': {u'field': <Products.PageTemplates.Expressions.PathIterator object at 0x134cebd0>},
       'nothing': None,
       'options': {'args': ()},
       'repeat': <Products.PageTemplates.Expressions.SafeMapping object at 0x14b1c260>,
       'request': <HTTPRequest, URL=http://localhost:8090/Plone/clients/client-1/AI15001/base_view>,
       'root': <Application at >,
       'template': <FSPageTemplate at /Plone/base_view used for /Plone/clients/client-1/AI15001>,
       'traverse_subpath': [],
       'user': <PropertiedUser 'admin'>}
  Module Products.PageTemplates.ZRPythonExpr, line 48, in __call__
   - __traceback_info__: context.widget(field.getName(), mode=mode, use_label=1)
  Module PythonExpr, line 1, in <expression>
  Module Products.Archetypes.BaseObject, line 287, in widget
  Module Products.Archetypes.Renderer, line 26, in render
  Module Products.Archetypes.generator.widget, line 147, in __call__
AttributeError: Macro datagridwidget does not exist for <ARImport at AI15001>
```

**Comment by rockfruit** [comment]
That should do it.  Meanwhile if you're not executing to319.py, you'll have to activate Products.DataGridField in site-setup/addons.

**Pull request merged**